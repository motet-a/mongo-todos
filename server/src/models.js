import V from '@motet-a/validate'

// Forget “models” as understood by ORMs and MVC, this files does one thing:
// it validates objects.
//
// Each “model” is a composable function-like object. In order to ensure
// that some value complies with the schema described by some “model”, simply
// call the “model” with the schema in parameter. It throws on error.
//
// See https://github.com/motet-a/validate for more information.

export const Task = V
  .shape({
    text: V
      .string
      .toTrimmed
      .max(1024),

    done: V
      .boolean
  })

export const TaskList = V
  .shape({
    _id: V
      .string,

    name: V
      .string
      .toTrimmed
      .max(80),

    tasks: V.array.of(Task)
  })

export default {Task, TaskList}
