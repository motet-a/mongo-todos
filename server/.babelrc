{
  "presets": [
    ["flow"], ["env"]
  ],
  "plugins": [
    ["babel-plugin-root-import", {"rootPathSuffix": "src"}]
  ]
}
