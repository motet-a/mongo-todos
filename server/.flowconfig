[ignore]

[include]

[libs]

[lints]

[options]
module.name_mapper='^{rootPathPrefix}/\(.*\)$' -> '<PROJECT_ROOT>{rootPathSuffix}/\1'

[strict]
