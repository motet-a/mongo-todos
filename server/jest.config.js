module.exports = {
  testEnvironment: './src/jest-environment',
  globals: {
    __TEST__: true
  }
}
