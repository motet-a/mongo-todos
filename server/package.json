{
  "name": "server",
  "private": true,
  "version": "1.0.0",
  "main": "index.js",
  "license": "MIT",
  "scripts": {
    "watch": "nodemon --exec babel-node src/index.js",
    "test": "jest",
    "lint": "standard --parser babel-eslint --global expect --global test --global describe --global beforeAll --global afterAll --global __TEST__ --global __DB__"
  },
  "dependencies": {
    "@motet-a/validate": "^0.1.1",
    "koa": "^2.4.1",
    "koa-bodyparser": "^4.2.0",
    "koa-logger": "^3.1.0",
    "koa-router": "^7.3.0",
    "koa-send": "^4.1.1",
    "lodash": "^4.17.4",
    "mongodb": "^2.2.33"
  },
  "devDependencies": {
    "babel-cli": "^6.26.0",
    "babel-core": "^6.26.0",
    "babel-eslint": "^8.0.2",
    "babel-jest": "^21.2.0",
    "babel-plugin-root-import": "^5.1.0",
    "babel-preset-env": "^1.6.1",
    "babel-preset-flow": "^6.23.0",
    "flow-bin": "^0.59.0",
    "jest": "21.3.0-beta.10",
    "node-fetch": "^1.7.3",
    "nodemon": "^1.12.1",
    "regenerator-runtime": "^0.11.0",
    "rollup": "^0.52.0",
    "standard": "^10.0.3"
  }
}
