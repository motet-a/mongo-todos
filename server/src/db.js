// Don't use ES6 imports here

const {MongoClient} = require('mongodb')

const url = process.env.TODO_MONGO_URL || 'mongodb://localhost:27017/todo'

const connect = async (options = {}) => {
  const db = await MongoClient.connect(url + (options.test ? '_test' : ''))
  global.__DB__ = db
  return db
}

const close = (...args) => {
  const result = __DB__.close(...args)
  global.__DB__ = null
  return result
}

const proxyfiedDb = new Proxy(
  {},

  {
    get (_target, name) {
      if (name === 'connect') {
        return connect
      }

      if (name === 'close') {
        return close
      }

      if (name === '__esModule') {
        return undefined
      }

      return global.__DB__[name]
    },

    set () {
      throw new Error()
    }
  }
)

module.exports = proxyfiedDb
