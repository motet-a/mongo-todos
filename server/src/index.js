global.__TEST__ = false

const db = require('./db')
const {default: app} = require('./koa-app')

const bootstrap = async () => {
  await db.connect()

  const port = 3000

  app.listen(3000, () => {
    console.log(`listening on port ${port}â€¦`)
  })
}

bootstrap()
  .catch(console.error)
