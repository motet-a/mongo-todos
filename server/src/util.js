import crypto from 'crypto'
import {promisify} from 'util'
import {ObjectID} from 'mongodb'

export async function secureRandomString (): string {
  const buffer = await promisify(crypto.randomBytes)(32)
  return buffer.toString('hex')
}

export function mongoId (): string {
  return (new ObjectID()).toHexString()
}

export default {secureRandomString, mongoId}
