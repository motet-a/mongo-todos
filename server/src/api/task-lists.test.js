import _ from 'lodash'

import server from '~/test-server'
import db from '~/db'
import models from '~/models'
import util from '~/util'

beforeAll(server.start)
afterAll(server.stop)

const taskListFixture = async () => {
  const fixture = {
    _id: await util.secureRandomString(),
    name: 'hello',
    tasks: []
  }

  models.TaskList(fixture)
  return fixture
}

describe('GET /task-lists/:id', () => {
  test('404 when the ID doesn’t exist', async () => {
    const res = await server.fetch('/api/task-lists/does-not-exist')
    expect(res.status).toEqual(404)
    expect(await res.text()).toEqual('Not Found')
  })

  test('200 ', async () => {
    const list = await taskListFixture()

    await db
      .collection('tasklists')
      .insertOne(list)

    const res = await server.fetch(`/api/task-lists/${list._id}`)
    expect(res.status).toEqual(200)
    expect(await res.json()).toEqual(list)
  })
})

describe('POST /task-lists', () => {
  test('201', async () => {
    const listToCreate = await taskListFixture()
    listToCreate.name = 'whitespace should be trimmed'

    const postData = Object.assign(
      {},
      listToCreate,
      {name: ` \n ${listToCreate.name} \t `}
    )

    const res = await server.fetchJson(`/api/task-lists`, {
      method: 'POST',
      body: postData
    })

    expect(res.status).toEqual(201)
    const createdList = await res.json()
    expect(_.omit(createdList, '_id')).toEqual(_.omit(listToCreate, '_id'))
    expect(createdList._id).not.toEqual(postData._id)
  })

  test('400', async () => {
    const res = await server.fetchJson('/api/task-lists', {
      method: 'POST',
      body: {}
    })
    expect(res.status).toEqual(400)
    expect(await res.json()).toMatchSnapshot()
  })
})

describe('DELETE /task-lists', () => {
  test('404 when the ID doesn’t exist', async () => {
    const res = await server.fetchJson(
      '/api/task-lists/does-not-exist',
      {method: 'DELETE'}
    )
    expect(res.status).toEqual(404)
  })

  test('204', async () => {
    const list = await taskListFixture()

    await db
      .collection('tasklists')
      .insertOne(list)

    const res = await server.fetchJson(
      `/api/task-lists/${list._id}`,
      {method: 'DELETE'}
    )
    expect(res.status).toEqual(204)
  })
})

describe('PUT /task-lists', () => {
  test('404 when the ID doesn’t exist', async () => {
    const res = await server.fetchJson(
      '/api/task-lists/does-not-exist',
      {method: 'PUT'}
    )
    expect(res.status).toEqual(404)
  })

  test('200 when an empty update occurs', async () => {
    const list = await taskListFixture()

    await db
      .collection('tasklists')
      .insertOne(list)

    const res = await server.fetchJson(
      `/api/task-lists/${list._id}`,
      {
        method: 'PUT',
        body: {}
      }
    )
    expect(res.status).toEqual(200)
    expect(await res.json()).toEqual(list)
  })

  test('200', async () => {
    const list = await taskListFixture()

    await db
      .collection('tasklists')
      .insertOne(list)

    const res = await server.fetchJson(
      `/api/task-lists/${list._id}`,
      {
        method: 'PUT',
        body: {name: 'new name'}
      }
    )
    expect(res.status).toEqual(200)

    const expectedNewList = Object.assign(
      {},
      list,
      {name: 'new name'}
    )
    expect(await res.json()).toEqual(expectedNewList)

    const insertedList = await db
      .collection('tasklists')
      .findOne({_id: list._id})

    expect(insertedList).toEqual(expectedNewList)
  })
})
