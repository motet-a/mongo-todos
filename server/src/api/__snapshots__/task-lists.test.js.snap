// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /task-lists 400 1`] = `
Object {
  "message": "Required value at \`name\`, got \`undefined\`",
  "path": Array [
    "name",
  ],
}
`;
